Added optional parameter `max` to replace functions

kintrix007 opened the pull request:
if `max` is `-1`, which is the default, it replaces all occurrences.
Can be optionally specified to only replace the first `max` amount.

I couldn't think of a nice, and relatively efficient way to implement this to `multiReplace`, while keeping it being one iteration... I could implement it using a table to track how many times it replaced given strings, or I could probably just implement a `multiReplacesFirsts` function.

kintrix007 committed:
Added optional parameter 'maxOccurrences' to replace functions

if 'maxOccurrences' is -1, which is the default, it replaces all occurrances. Can be optionally specified to only replace first 'maxOccurrences' ones. I couldn't think of a nice, and efficient way to implement this in multiReplace, to keep it being one iteration... I could implement it using a table to track how many times it replaced given strings, or I could just implement a 'multiReplacesFirsts' function

kintrix007 committed:
Commented unnecessary function

kintrix007 committed:
Added 'runnableExamples', small hotfix

timotheecour commented on lib/pure/strutils.nim:
```diff
@@ -2050,97 +2050,145 @@ func contains*(s: string, chars: set[char]): bool =
+    extern: "nsuReplaceChar".} =
+  ## Replaces `maxOccurrence` occurrences of the character `sub` in `s` with
+  ## the character `by`.
+  ## If `maxOccurrence` is set to `-1`, it replaces all occurrences.
```
```suggestion
  ## If `maxOccurrence == -1`, it replaces all occurrences.
```
(shorter = better)

ditto elsewhere

timotheecour commented on lib/pure/strutils.nim:
```diff
@@ -2050,97 +2050,145 @@ func contains*(s: string, chars: set[char]): bool =
+  ## * `replaceWord func<#replaceWord,string,string,string>`_
+  ## * `multiReplace func<#multiReplace,string,varargs[]>`_
+  runnableExamples:
+    doAssert "valid variable name".replace(' ', '_') == "valid_variable_name"
```
* use assert in runnableExamples (and doAssert in all other tests), see contributing.html

kintrix007 replied:
In many places `doAssert` was already used in `runnableExamples`, which weren't made by me.
Then it would make sense to have some PR's that change all...?

timotheecour replied:
new code should stick to the currently accepted best practice, not surrounding code; and yes, a PR that changes all those should be done (and in fact i wrote it here: https://github.com/nim-lang/Nim/pull/17616 but i still need to undraft it even though i think it's ready)

timotheecour commented on lib/pure/strutils.nim:
```diff
@@ -2050,97 +2050,145 @@ func contains*(s: string, chars: set[char]): bool =
+  ## the character `by`.
+  ## If `maxOccurrence` is set to `-1`, it replaces all occurrences.
+  ##
+  ## Optimized version of `replace <#replace,string,string,string>`_ for
```
add corresponding tests to tstrutils (so that it's also tested in VM + js)

timotheecour commented on lib/pure/strutils.nim:
```diff
@@ -2050,97 +2050,145 @@ func contains*(s: string, chars: set[char]): bool =
   return find(s, chars) >= 0
 
-func replace*(s, sub: string, by = ""): string {.rtl,
+func replace*(s: string, sub, by: char, maxOccurrence = -1): string {.rtl,
```
`s/maxOccurrence/max/`  (maxOccurrence is too long to type/remember when used as optional param)

and max is just as clear

timotheecour commented on lib/pure/strutils.nim:
```diff
@@ -2050,97 +2050,145 @@ func contains*(s: string, chars: set[char]): bool =
   ## * `replaceWord func<#replaceWord,string,string,string>`_
   ## * `multiReplace func<#multiReplace,string,varargs[]>`_
+  runnableExamples:
+    doAssert "This is not {}, but can be used as {}".replace("{}", "fmt()") == "This is not fmt(), but can be used as fmt()"
```
prefer simple illustrative examples (including edge cases) to "cute" but more complex examples

kintrix007 committed:
Update lib/pure/strutils.nim

Co-authored-by: Timothee Cour <timothee.cour2@gmail.com>

kintrix007 committed:
added tests, changed  to  in runnableExamples, added it to changelog.md

kintrix007 committed:
fixed conflict

timotheecour commented on lib/pure/strutils.nim:
```diff
@@ -2114,8 +2114,8 @@ func replace*(s, sub: string, by = "", maxOccurrence = -1): string {.rtl,
   runnableExamples:
-    doAssert "This is not {}, but can be used as {}".replace("{}", "fmt()") == "This is not fmt(), but can be used as fmt()"
-    doAssert "almost valid variable name".replace(" ", "_", 2) == "almost_valid_variable name"
+    assert "FooBarBaz".replace("Foo", "Bar") == "BarBarBaz"
```
this doesn't illustrate that it replaces all occurrences since there's a single occurrence

kintrix007 replied:
That is a stupid oversight, sorry

kintrix007 committed:
improved runnableExamples

kintrix007 committed:
added more edge-case checks, which made sense

kaushalmodi commented on lib/pure/strutils.nim:
```diff
@@ -2050,97 +2050,145 @@ func contains*(s: string, chars: set[char]): bool =
+    extern: "nsuReplaceChar".} =
+  ## Replaces `max` occurrences of the character `sub` in `s` with
+  ## the character `by`.
+  ## If `maxOccurrence == -1`, it replaces all occurrences.
```
The parameter name is not `maxOccurrence`. Need to just replace that with `max`.

timotheecour replied:
```suggestion
  ## If `max == -1`, it replaces all occurrences.
```

timotheecour committed:
Update lib/pure/strutils.nim

timotheecour commented on lib/pure/strutils.nim:
```diff
@@ -2050,97 +2050,145 @@ func contains*(s: string, chars: set[char]): bool =
+  result = newString(s.len)
+  var occLeft = max
+  for i in 0..s.high:
+    if occLeft != 0 and s[i] == sub:
```
this is inefficient; occLeft == 0, we should break

ditto with other procs

kintrix007 replied:
I think the other ones already did break...
I didn't break here, since we needed to add the rest of the string either way. But now changed it.

kintrix007 committed:
did not use  correctly, changed tests to catch that, and fixed the bug

kintrix007 committed:
Merge branch 'strutils-replace-improvements' of https://github.com/kintrix007/Nim into strutils-replace-improvements

kintrix007 committed:
'replaceWord' did not use 'max' correctly, changed tests to catch that, and fixed the bug

kintrix007 committed:
'replaceWord' did not use 'max' correctly, changed tests to catch that, and fixed the bug

kaushalmodi commented on lib/pure/strutils.nim:
```diff
@@ -2050,97 +2050,152 @@ func contains*(s: string, chars: set[char]): bool =
+    # copy the rest:
+    result.add s.substr(i)
+
+# Maybe implement something like this for all replace functions...?
```
@kintrix007 Why do we have this commented block here? Should this instead be a separate issue or RFC?

kintrix007 replied:
We do not.
I left it in case someone had something to say about it.
Will definitely remove it if the PR gets accepted.
But now that you mention it, I should have probably just said that as a comment here...

kaushalmodi replied:
> Will definitely remove it if the PR gets accepted.

If the PR gets accepted, someone will just merge it. So it's better to keep the PR in the state you want to see it merged.

timotheecour replied:
also, the implementation is not good (no early break), so if you want to keep a comment just add:
```nim
# consider adding: func replace*(s: string, subs: set[char], by: char, max = -1): string 
```
(whether it's a good idea or not, is unclear)

timotheecour commented on lib/pure/strutils.nim:
```diff
@@ -2050,97 +2050,152 @@ func contains*(s: string, chars: set[char]): bool =
+      let j = find(a, s, sub, i)
+      if j == -1 or occLeft == 0: break
+      if occLeft > 0: dec occLeft
+      result.add s.substr(i, j - 1)
```
pre-existing but this creates un-necessary allocations.

IMO we need a 0-allocation [1] `addReplace` in strbasics and then strutils.replace can reuse it.

not a blocker for this PR though as this refactoring can be done later.

```nim
proc addReplace*(result: var string, input: openArray[char], pattern: string, by: string, max = -1): int
  ## returns the number of `pattern` instances found in `input` and replaced by `by`,
  ## the result being appended to `result`
```

(0-alloc because user can reuse the result buffer)

Araq replied:
We should use a TR macro to optimize this, the pattern is common and while we can avoid it, many of our users don't know about it.

timotheecour replied:
that won't make `replace` 0-alloc though, hence the need for `strbasics.addReplace` in future work.

Note that for that particular line, all that's needed is replace `substr` by `toOpenArray`:
```
      result.add s.substr(i, j - 1)
=>
      result.add s.toOpenArray(i, j - 1)
```
(with `import std/strbasics` on top), which is now possible since #15951
regarding TR here, maybe (but TR have their own issues)

kintrix007 committed:
Merge branch 'nim-lang:devel' into strutils-replace-improvements

kintrix007 committed:
changed replace overload suggestion to single line comment, added another test case

kintrix007 committed:
imported std/strbasics

tests fail after importing std/strbasics, and using toOpenArray, the code compiles and doc still succeeds

kintrix007 committed:
Merge branch 'nim-lang:devel' into strutils-replace-improvements

timotheecour commented:
@kintrix007 the diff shows unrelated changes; also, I recommend to embrace git rebase instead of git merge (easier to follow PR's)

kintrix007 commented:
> @kintrix007 the diff shows unrelated changes;

I have no idea how those happened, as those are under `bin/`, which is in the gitignore file
